### ISSUE TITLE
Chose compiler service

### ISSUE THREAD

**Issue by Leandros** [cpp-tour/cpp-tour#1]
Current alternatives:

- ~[Coliru](http://coliru.stacked-crooked.com/) (used by cppreference)~
- [Wandbox](https://github.com/melpon/wandbox/blob/master/kennel2/API.rst)
- [x86 OS emulator in JS?](https://copy.sh/v86/?profile=linux3)

Decided on Wandbox!

**Comment by grisumbras** [comment]
I like Wandbox'es API more. Feature-wise they seem mostly equivalent, but Wandbox supports using multiple files (can be useful, for more advanced lessons), and returns result as  JSON (probably easier to work with).
Coliru, on the other hand, is slightly faster (on my machine).

**Comment by Leandros** [comment]
Agreed! I think the choice of Wandbox to start with is obvious. Let's go with that.

**Comment by DoumanAsh** [comment]
@Leandros What should we expect for you all editor UI aside from compile button and output window?
Should we have way to provide expected output for our lessons?

**Comment by Leandros** [comment]
My current vision for the editor UI is to have two buttons: "Reset" and "Run". Reset will reset the code to the original code snippet, while "Run" will compile and run the code.
We should probably also show a tab with the current "filename", to get users used to seeing multiple files in an editor window?

**Comment by graph** [comment]
I'm too late too late to the party. Maybe not too late to change.

How about a client side compiler? what about emscripten? or getting a client side virtual linux box?  Online service have queues, and sometimes it takes a while to compile, plus there is internet component. Emscripten these days is very fast, and now wasm support is very popular. And there is also x86 emulators in javascript which can run linux all using client side resources. Lots of options how to set up the compiler on client side. With the x86 emulator approach you can continuously run cppcheck or the like as user types and be more speedy. webworkers on the web can do it in background.

I don't know how emscripten would work. I think the only client side option is one of the x86 emulators in javascript and setup linux with gcc & clang setup.

**Comment by DoumanAsh** [comment]
If you're aware of existing solution that would let us to complie C++ to wasm, I'm all for that.

But I'm not sure if there any solution such as it at the moment, so for now we can go with wandbox API.

**Comment by graph** [comment]
https://copy.sh/v86/?profile=linux3

I'm sure gcc/clang can be installed some how. And the download images to be cached so it's faster for repeat visitors. It's definitely more work to setup than wandbox.

**Comment by DoumanAsh** [comment]
@graph It seems to be quite surprising to run the whole OS in JS.
Will need to investigate it further so for now I'll pin it to first post.
Thanks

**Comment by graph** [comment]
Thanks for investigating. This route is alot of work so I can understand not going this route. And this project would be the leader of client side c++ compiling as I don't think a virtual machine has been setup for this yet. If your team does go this route I think it would be very cool. When I use  some site that uses a compile service, the queues can get quite long. Something client side will definitely help with that.

**Comment by hermanzdosilovic** [comment]
Maybe you could use my free service [Judge0 API](https://api.judge0.com). It's available on [GitHub](https://github.com/judge0/api) and you can try running some C++ code in my online code editor [Judge0 IDE](https://ide.judge0.com). Let me know how can I help you if you choose to use Judge0.
